File: 01-coding/python/exercises/health_checker.py
#!/usr/bin/env python3
"""
Service Health Checker (Exercise)

OBJECTIVE:
Build a robust HTTP health checker with retry logic and exponential backoff.

REQUIREMENTS:
1. Create a HealthChecker class that:
   - Takes URL, timeout, and max_retries as parameters
   - Performs HTTP GET health checks
   - Implements retry logic with exponential backoff (2^attempt seconds)
   - Returns detailed health check results

2. Health check should return a dictionary with:
   - url: The URL checked
   - status_code: HTTP status code
   - response_time: Time taken for the request (in seconds)
   - healthy: Boolean (True if status code is 2xx)
   - attempt: Which attempt succeeded (or total attempts if failed)

3. Retry behavior:
   - Retry on timeout, connection errors, or non-2xx responses
   - Wait 2^attempt seconds between retries (2s, 4s, 8s, etc.)
   - After all retries fail, return unhealthy status with error message

4. Use structured logging to track:
   - Each attempt and its outcome
   - Wait times between retries
   - Success or failure status

5. Accept command-line arguments:
   - url (required): URL to check
   - --timeout: Request timeout in seconds (default: 5)
   - --retries: Number of retry attempts (default: 3)

6. Exit with code 0 if healthy, 1 if unhealthy (for monitoring integration)

HINTS:
- Use requests library (pip install requests)
- Use time.sleep() for waits
- Use time.time() to measure response time
- Catch requests.exceptions.Timeout and requests.exceptions.ConnectionError
- Use argparse for command-line arguments
- Use logging module for structured logging

TODO: Implement the health checker below
"""

import sys
import time
import logging
import argparse
from typing import Dict, Optional

# TODO: Import requests library and handle ImportError if not installed


# TODO: Configure logging


class HealthChecker:
    """Service health checker with retry logic."""
    
    def __init__(self, url: str, timeout: int = 5, max_retries: int = 3):
        # TODO: Initialize instance variables
        pass
    
    def check_health(self) -> Dict[str, any]:
        """
        Check service health with retry logic.
        
        Returns:
            dict: Health check results including status, response time, etc.
        """
        # TODO: Implement health check with retry logic
        # - Loop through attempts (1 to max_retries)
        # - Make HTTP GET request with timeout
        # - Measure response time
        # - Check if status code is 2xx (healthy)
        # - If healthy, return success result
        # - If unhealthy or error, log and retry after exponential backoff
        # - Return failure result if all retries exhausted
        pass


def main():
    # TODO: Implement main function
    # - Parse command-line arguments (url, timeout, retries)
    # - Create HealthChecker instance
    # - Call check_health()
    # - Display results
    # - Return 0 if healthy, 1 if unhealthy
    pass


if __name__ == "__main__":
    # TODO: Call main() and exit with its return value
    pass

File: 01-coding/golang/exercises/health_checker.go
package main

/*
HTTP Health Checker - Go Edition (Exercise)

OBJECTIVE:
Build a concurrent health checker that can check multiple URLs simultaneously.

REQUIREMENTS:
1. Create a HealthCheck struct with fields:
   - URL (string)
   - StatusCode (int)
   - ResponseTime (time.Duration)
   - Healthy (bool)
   - Error (error)

2. Implement CheckHealth() function:
   - Takes context and URL as parameters
   - Makes HTTP GET request with 5-second timeout
   - Returns HealthCheck struct with results
   - Mark as healthy if status code is 2xx
   - Measure response time

3. Implement CheckMultipleServices() function:
   - Takes slice of URLs
   - Checks all URLs concurrently using goroutines
   - Uses channels to collect results
   - Returns slice of HealthCheck results

4. Main function should:
   - Define default URLs or accept them as command-line arguments
   - Call CheckMultipleServices()
   - Display results for each service
   - Show summary (X/Y services healthy)
   - Exit with code 1 if any service is unhealthy

HINTS:
- Use net/http package for HTTP requests
- Use context.WithTimeout() for request timeouts
- Use goroutines for concurrent checks: go func() { ... }()
- Use channels to collect results: make(chan HealthCheck, len(urls))
- Use time.Since() to measure duration
- Use http.NewRequestWithContext() to create requests with context
- Check status code: resp.StatusCode >= 200 && resp.StatusCode < 300
- Don't forget to close response body: defer resp.Body.Close()

TODO: Implement the program below
*/

import (
	// TODO: Import required packages:
	// - context
	// - fmt
	// - net/http
	// - os
	// - strings
	// - time
)

// TODO: Define HealthCheck struct


// TODO: Implement CheckHealth function
// func CheckHealth(ctx context.Context, url string) HealthCheck


// TODO: Implement CheckMultipleServices function
// func CheckMultipleServices(urls []string) []HealthCheck


func main() {
	// TODO: Define default URLs (or get from command-line arguments)
	
	
	// TODO: Display header
	
	
	// TODO: Call CheckMultipleServices and get results
	
	
	// TODO: Display results for each service
	// Show status (✅ HEALTHY or ❌ UNHEALTHY), URL, status code, response time
	
	
	// TODO: Display summary showing healthy count vs total
	
	
	// TODO: Exit with code 1 if any service is unhealthy
	
}
